src/cargo/core/workspace.rs

```diff
@@ -991,11 +980,38 @@ impl<'cfg> Workspace<'cfg> {
                 if !manifest.patch().is_empty() {
                     emit_warning("patch")?;
                 }
-                if let Some(behavior) = manifest.resolve_behavior() {
-                    if behavior != self.resolve_behavior {
-                        // Only warn if they don't match.
-                        emit_warning("resolver")?;
+                if manifest.defaulted_resolve_behavior() != self.resolve_behavior {
+                    let package_reason = manifest
+                        .resolve_behavior()
+                        .is_none()
+                        .then(|| format!(r#" implied by edition = "{}""#, manifest.edition()))
+                        .unwrap_or_default();
+                    let workspace_reason = match self.root_maybe() {
+                        MaybePackage::Package(p) => {
+                            p.manifest().resolve_behavior().is_none().then(|| {
+                                format!(r#" implied by edition = "{}""#, manifest.edition())
+                            })
+                        }
+                        MaybePackage::Virtual(vm) => {
+                            vm.resolve_behavior().is_none().then(|| " (default)".into())
+                        }
                     }
+                    .unwrap_or_default();
+                    let msg = format!(
+                        "expected resolver for package is overridden by the workspace root:\n\
+                         package:   {} expects resolver = \"{}\"{package_reason}\n\
+                         workspace: {} sets resolver = \"{}\"{workspace_reason}",
+                        pkg.manifest_path().display(),
+                        manifest
+                            .defaulted_resolve_behavior()
+                            .to_manifest()
+                            .unwrap_or_else(|| "1".into()),
+                        root_manifest.display(),
+                        self.resolve_behavior
+                            .to_manifest()
+                            .unwrap_or_else(|| "1".into()),
+                    );
+                    self.config.shell().warn(&msg)?;
```

Comment:
Emitting the message per package will be very noisy in workspaces.  

My recommendation:
- We leave the existing warning as-is
- We add a new warning similar to what I was in the middle of writing :)

```rust
let members_resolvers = self
    .members()
    .map(|p| p.manifest().resolve_behavior())
    .collect::<HashSet<_>>();
let shared_resolver = members_resolvers.len() == 1;
let member_resolver = members_resolvers.iter().next().copied().flatten();
if shared_resolver && member_resolver != Some(ResolveBehavior::V1) {
    let member_resolver = member_resolver.unwrap();
    self.config.shell().warn(format!(
        "applying default workspace.resolver=1 despite members having package.resolver={}.",
        member_resolver.to_manifest().unwrap(),
    ))?;
}
```

Reply:
If it's silenced by explicitly setting the resolver then that will still lead to inconsistent workspaces. Even if it's set to `workspace.resolver = "2"` then in the future we may have `package.edition = "2030"` imply resolver `"3"` but there will be no warning emitted about the package using different resolution in the workspace vs. when published.

Reply:
> workspace vs. when published

Normally when I see this problem being discussed, its from the angle of local build failures.  Sounds like you are concerned about build differences between local and published which is mostly just docs.rs and `cargo install`.  Is that correct?  Could you note that on the original issue and let's continue the discussion there.